### PR TITLE
fix: multitenancy passwordless implementation

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

- `tenantId` in email and sms delivery
- emailverification implementation
- passwordless implementation
- thirdparty implementation

## Related issues
https://github.com/supertokens/supertokens-node/pull/471

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [ ] Had run `npm run build-pretty`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

## Remaining TODOs for this PR

-   [ ] Item1
-   [ ] Item2

### CODE REVIEW

File: lib/ts/recipe/passwordless/api/implementation.ts [hunk 1]
```diff
@@ -167,6 +175,7 @@ export default function getAPIImplementation(): APIInterface {
             let deviceInfo = await input.options.recipeImplementation.listCodesByDeviceId({
                 userContext: input.userContext,
                 deviceId: input.deviceId,
+                tenantId: input.tenantId,
```
Comment (rishabhpoddar) [inline lib/ts/recipe/passwordless/api/implementation.ts:178]: write a test case which tries to call this API with a different tenantId than the one the flow was initiated with and make sure that we get back RESTART_FLOW_ERROR

File: lib/ts/recipe/passwordless/api/implementation.ts [hunk 2]
```diff
@@ -97,6 +101,8 @@ export default function getAPIImplementation(): APIInterface {
                     input.options.recipeId +
                     "&preAuthSessionId=" +
                     response.preAuthSessionId +
+                    "&tenantId=" +
+                    input.tenantId +
                     "#" +
```
Comment (rishabhpoddar) [inline lib/ts/recipe/passwordless/api/implementation.ts:106]: pass in tenantId to sendSms and sendEmail functions as well.

File: lib/ts/recipe/passwordless/api/implementation.ts [hunk 3]
```diff
@@ -97,6 +101,8 @@ export default function getAPIImplementation(): APIInterface {
...
                     response.linkCode;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/passwordless/api/implementation.ts:107]: pass in tenantId to sendEmail in emailpassword and in email verification recipes as well.

File: lib/ts/recipe/passwordless/recipe.ts [hunk 4]
```diff
@@ -237,11 +239,13 @@ export default class Recipe extends RecipeModule {
...
+                      tenantId: input.tenantId,
                       userContext: input.userContext,
                   }
                 : {
                       phoneNumber: input.phoneNumber,
                       userInputCode,
+                      tenantId: input.tenantId,
                       userContext: input.userContext,
                   }
         );
```
Comment (rishabhpoddar) [inline lib/ts/recipe/passwordless/recipe.ts:251]: need to add tenantId in the generated magic link below

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 5]
```diff
@@ -495,6 +518,7 @@ export type APIInterface = {
         | undefined
         | ((input: {
               phoneNumber: string;
+              tenantId: string;
               options: PasswordlessAPIOptions;
               userContext: any;
           }) => Promise<
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:524]: TypeThirdPartyPasswordlessEmailDeliveryInput and TypeThirdPartyPasswordlessSmsDeliveryInput should take tenantId

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/thirdpartypasswordless/types.ts (hunk 5 above)
Comment (rishabhpoddar) [reply lib/ts/recipe/thirdpartypasswordless/types.ts:524]: Same goes for in the passwordless / emailpassword / emailverification recipes

File: lib/ts/recipe/passwordless/api/implementation.ts (hunk 2 above)
Comment (sattvikc) [reply lib/ts/recipe/passwordless/api/implementation.ts:106]: Done

File: lib/ts/recipe/passwordless/api/implementation.ts (hunk 3 above)
Comment (sattvikc) [reply lib/ts/recipe/passwordless/api/implementation.ts:107]: Done

File: lib/ts/recipe/passwordless/api/implementation.ts (hunk 1 above)
Comment (sattvikc) [reply lib/ts/recipe/passwordless/api/implementation.ts:182]: Done

File: lib/ts/recipe/passwordless/recipe.ts (hunk 4 above)
Comment (sattvikc) [reply lib/ts/recipe/passwordless/recipe.ts:251]: Done

File: lib/ts/recipe/thirdpartypasswordless/types.ts (hunk 5 above)
Comment (sattvikc) [reply lib/ts/recipe/thirdpartypasswordless/types.ts:526]: Done

File: lib/ts/recipe/thirdpartypasswordless/index.ts [hunk 6]
```diff
@@ -249,17 +249,21 @@ export default class Wrapper {
         });
     }
 
-    static async sendEmail(input: TypeThirdPartyPasswordlessEmailDeliveryInput & { userContext?: any }) {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/index.ts:202]: how come thirdPartyManuallyCreateOrUpdateUser is compiling without tenantId as an input?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/emailpassword/index.ts [hunk 7]
```diff
@@ -86,11 +86,12 @@ export default class Wrapper {
         });
     }
 
-    static async sendEmail(input: TypeEmailPasswordEmailDeliveryInput & { userContext?: any }) {
+    static async sendEmail(input: TypeEmailPasswordEmailDeliveryInput & { tenantId?: string; userContext?: any }) {
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/emailpassword/index.ts:89]: Done

File: lib/ts/recipe/emailverification/utils.ts [hunk 8]
```diff
@@ -69,14 +70,21 @@ export function validateAndNormaliseUserInput(
     };
 }
 
-export function getEmailVerifyLink(input: { appInfo: NormalisedAppinfo; token: string; recipeId: string }): string {
+export function getEmailVerifyLink(input: {
+    appInfo: NormalisedAppinfo;
+    token: string;
+    recipeId: string;
+    tenantId?: string;
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/emailverification/utils.ts:77]: Done

File: lib/ts/recipe/emailverification/utils.ts [hunk 9]
```diff
@@ -69,14 +70,21 @@ export function validateAndNormaliseUserInput(
...
         input.appInfo.websiteDomain.getAsStringDangerous() +
         input.appInfo.websiteBasePath.getAsStringDangerous() +
         "/verify-email" +
         "?token=" +
         input.token +
         "&rid=" +
-        input.recipeId
+        input.recipeId +
+        "&tenantId=" +
+        (input.tenantId === undefined ? DEFAULT_TENANT_ID : input.tenantId)
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/emailverification/utils.ts:88]: Done

File: lib/ts/recipe/passwordless/index.ts [hunk 10]
```diff
@@ -212,17 +212,19 @@ export default class Wrapper {
...
+    static async sendEmail(input: TypePasswordlessEmailDeliveryInput & { tenantId?: string; userContext?: any }) {
         return await Recipe.getInstanceOrThrowError().emailDelivery.ingredientInterfaceImpl.sendEmail({
             userContext: {},
             ...input,
+            tenantId: input.tenantId === undefined ? DEFAULT_TENANT_ID : input.tenantId,
         });
     }
 
-    static async sendSms(input: TypePasswordlessSmsDeliveryInput & { userContext?: any }) {
+    static async sendSms(input: TypePasswordlessSmsDeliveryInput & { tenantId?: string; userContext?: any }) {
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/passwordless/index.ts:223]: Done

File: lib/ts/recipe/emailverification/api/implementation.ts [hunk 11]
```diff
@@ -142,6 +145,7 @@ export default function getAPIInterface(): APIInterface {
                         email: emailInfo.email,
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/emailverification/api/implementation.ts:145]: Done

File: lib/ts/recipe/emailverification/index.ts [hunk 12]
```diff
@@ -137,11 +142,12 @@ export default class Wrapper {
         });
     }
 
-    static async sendEmail(input: TypeEmailVerificationEmailDeliveryInput & { userContext?: any }) {
+    static async sendEmail(input: TypeEmailVerificationEmailDeliveryInput & { tenantId?: string; userContext?: any }) {
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/emailverification/index.ts:145]: Done

File: lib/ts/recipe/thirdpartyemailpassword/index.ts [hunk 13]
```diff
@@ -109,10 +110,11 @@ export default class Wrapper {
         });
     }
 
-    static async sendEmail(input: TypeEmailPasswordEmailDeliveryInput & { userContext?: any }) {
+    static async sendEmail(input: TypeEmailPasswordEmailDeliveryInput & { tenantId?: string; userContext?: any }) {
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdpartyemailpassword/index.ts:113]: Done

File: lib/ts/recipe/thirdpartypasswordless/index.ts [hunk 14]
```diff
@@ -249,17 +249,21 @@ export default class Wrapper {
...
+    static async sendEmail(
+        input: TypeThirdPartyPasswordlessEmailDeliveryInput & { tenantId?: string; userContext?: any }
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdpartypasswordless/index.ts:253]: Done

File: lib/ts/recipe/thirdpartypasswordless/index.ts [hunk 15]
```diff
@@ -249,17 +249,21 @@ export default class Wrapper {
...
+    ) {
         return await Recipe.getInstanceOrThrowError().emailDelivery.ingredientInterfaceImpl.sendEmail({
             userContext: {},
             ...input,
+            tenantId: input.tenantId === undefined ? DEFAULT_TENANT_ID : input.tenantId,
         });
     }
 
-    static async sendSms(input: TypePasswordlessSmsDeliveryInput & { userContext?: any }) {
+    static async sendSms(input: TypePasswordlessSmsDeliveryInput & { tenantId?: string; userContext?: any }) {
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/thirdpartypasswordless/index.ts:262]: Done

File: lib/ts/recipe/thirdparty/api/implementation.ts [hunk 16]
```diff
@@ -54,7 +54,16 @@ export default function getAPIInterface(): APIInterface {
                 thirdPartyUserId: userInfo.thirdPartyUserId,
                 email: emailInfo.id,
                 oAuthTokens: oAuthTokensToUse,
-                rawUserInfoFromProvider: userInfo.rawUserInfoFromProvider,
+                rawUserInfoFromProvider: {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdparty/api/implementation.ts:57]: make fromIdTokenPayload and fromUserInfoAPI optional everywhere.

File: lib/ts/recipe/thirdparty/api/implementation.ts [hunk 17]
```diff
@@ -97,7 +106,16 @@ export default function getAPIInterface(): APIInterface {
                 user: response.user,
                 session,
                 oAuthTokens: oAuthTokensToUse,
-                rawUserInfoFromProvider: userInfo.rawUserInfoFromProvider,
+                rawUserInfoFromProvider: {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdparty/api/implementation.ts:109]: this normalisation not required

File: lib/ts/recipe/thirdparty/api/implementation.ts [hunk 18]
```diff
@@ -54,7 +54,16 @@ export default function getAPIInterface(): APIInterface {
                 thirdPartyUserId: userInfo.thirdPartyUserId,
                 email: emailInfo.id,
                 oAuthTokens: oAuthTokensToUse,
-                rawUserInfoFromProvider: userInfo.rawUserInfoFromProvider,
+                rawUserInfoFromProvider: {
+                    fromIdTokenPayload:
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdparty/api/implementation.ts:58]: this normalisation not required

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**